### DISCUSSION TITLE
Improve Rack::Static

### DISCUSSION THREAD

**Discussion by foonlyboy** [rack/rack#2061] [Ideas]
Hi at the rack

I'd like to enhance Rack::Static

I was looking for a way to easily serve static files in rails development.
This can easily be done with Rack::Static

But Rack::Static is very simplistic.
It only allows for a path from the url to be appended to a root path in the file system.
I'd like to improve on that.

 My idea is to have a Hash of mappings: {}
- String    => String: should behave as before, append the url path to root path
- Regexp => String: should match on regexp, and then apply gsub(regexp, string) (FIXME: avoid bad paths)
- * => Proc should execute Proc on mathc
- Proc => *: should match Proc on true
- Proc => Proc should allow the full freedom

Hint: in rails config/application.rb add:
    config.middleware.insert_after(
      ActionDispatch::Static,
      Rack::Static, urls: ['static'], root: '/some-path/'
    )

**Comment by ioquatix** [comment]
I think in most cases, it's better to have simple middleware that layer on top of each other.

Can you explain the problem you are thinking about solving? i.e. why do you want to make this improvement.

**Comment by foonlyboy** [comment]
Hi Samuel,
thanks for your quick reply

I fully agree with you,
but then look into Rack::Static
there is a lot of cruft in that.
And the same applies to Rack::Files

I'm just new with Rack,
but I already got it, that it is not yet stackable?

As far as I have learned,
the rack modules are only executed sequentially?

Maybe I'm heading for xrack?
just kidding ...

let's talk about that
Best regards,
Eike Dierks

**Reply by ioquatix** [reply]
If you have any thoughts about simplifying the existing code or even just extracting some code into a method which makes it easier to sub-class, both would be okay by me, we'd need to consider the specific changes on their own merits though.

**Comment by jeremyevans** [comment]
I wouldn't want to maintain all of this extra complexity in rack.  You may want to copy the Rack::Static code and modify it to add the features you want, and then ship that in a gem for others to use.

**Comment by foonlyboy** [comment]
HI Samuel

I came here from a simple problem.
I'd like to explain the problem and my suggested solution.

I have an app.
The urls are based on /<path>

I want to rebase the app.
I want them to be based on /my-app/<path>

I got it with the rails routes  for the app
but then also my static files need new routes.

I had a custom solution for that,
but then I learned, about Rack::Static
(which obviously is the road to go)

But here is the problem:
Rack::Static simply concats the paths:
  /myapp/path resolves to /myapp/myapp/path

I'd want to come up with some new code.
I'm asking you to review it.

I'd like to suggest an extension to the api,
but that should be compatible with all legacy code.

Rack::Static.new(mappings: {})

{String => String} would behave as before.
{RegExp => String} would match on RegExp and route to gsub(RegExp, String)

But then there's even more:
{Proc => Proc} opens up all the routing, at your full disposal


I fully understand that this is some new complication.
But then look at the code. How does this interact with ::Files?
This looks to me like a bad hack.

--
I have some new ideas for the rack.
The rack should be stackable.
I'm quite new at the rack.


Best regards,
Eike Dierks

**Reply by jeremyevans** [reply]
Personally, I think Rack::Static is already a bit complicated.  I'm not really in favor of proposals to make it even more complicated, especially proposals where the benefit of making it more complicated seem fairly minimal.

I don't think Rack::Static's integration with Rack::Files is a hack.  That part works fairly well, IMO, especially how it can call Rack::Files multiple times in order to serve gzipped static files for better performance.  Maybe I'm biased as I wrote that part.

I don't want to discourage you.  If you submit a pull request with the proposed changes, we will consider it.  However, know upfront there is a high likelihood that it would not be accepted, and be prepared for that if you do submit such a pull request.  You have a higher chance of acceptance the simpler the patch is.  If we don't accept it, you can still distribute it yourself.

One possibility that wouldn't be too complicated would be to support a block:

```ruby
  use Rack::Static, urls: ['/myapp/static'], :root=>'public' do |path|
    # request path is yielded, modified path should be returned
  end
```

I'm not sure we would even accept that, but the more complex it gets, the less likely it is to be accepted.

**Comment by matthewd** [comment]
I think you're looking for `Rack::URLMap`, either to mount your entire existing application, or possibly combined with `Rack::Cascade`.

You might be able to use something like:

```ruby
class Intercept < Rack::Cascade
  def initialize(main_app, injected_app)
    super([injected_app, main_app])
  end
end

# ...

use Intercept, Rack::URLMap("/myapp" => Rack::Static.new(...))
```

**Comment by foonlyboy** [comment]
Hi Jeremy,

I do not want to change Rack::Static.
This would be too complicated to stay compatible.

But you got my idea.
~eike

**Comment by foonlyboy** [comment]
Hi Jeremy,
I have to look at that.

Back then, the rack was very easy.
We should really keep the basic rack like that.

But then, rack has the ability to stack modules.
But up to now this is only sequential.

I'd like to suggest that we stay sequential with the basic rack api,
but to introduce xrack as a new api.

I don't even know why we should need that.
But I had problems rewriting the requests with basic rack.

I would imagine, that we can have routers in the rack chain,
to divert the routes to different handlers.

I come up with a nice route filter,
this is something that I want to show to you.

~eike

**Comment by rubyFeedback** [comment]
I am trying to move away from my legacy .cgi files to rack.

Unfortunately lack of documentation makes this almost impossible.

For instance, I need to be able to serve my .css files from one directory; but
my local images are spread over multiple directories.

I read that I can use:

    use Rack::Static, urls: ["/images"], root: "/home/x/images"

But, finding documentation about this that is meant to read for simple humans,
is mission impossible.

I am not using rackup from the commandline, but 95% of the tutorials say to use rackup. I am using
e. g.

    Rackup::Server.start(
      app: Rack::ShowExceptions.new(Web::Actions::RackController.new),
           Port: Web::Actions::RackController::PORT_TO_USE
    )

The problem is: where to next put "use Rack::Static"? Putting it into my
custom class RackController does not work; rack/ruby says it can not find
use. Ok. I grepped the sources; turns out use is part of Rack::Builder. 
But I am not using it right now. So how can I tell Rack where my local
directories are? Why do I have to use "use" to begin with? Are there no
toplevel methods available such as Rack.set_local_directory() or 
something like that? This everything-must-be-a-DSL-but-nobody-writes-documentation
is really killing ruby. Just have a look and try to find a good tutorial about
rack; I can find useful snippets, and store them locally, but there is barely
any coherency in this, and many key things I still can not find out.

In the past, I had a similar impression about sinatra - cool for simple stuff
but once it gets more complicated, the documentation will bite you down
into surrender-mode. Now I realise that this problem already originates from 
rack itself. 

Anyway. I upvoted foonlyboy's suggestion - not necessarily because I think
we need more complexity, but because I myself want a simple way to 
specify different directories for images. I have one primary directory
at /home/x/data/images/, but also other images elsewhere on my local
filesystem under linux. Right now I have no idea how to tell rack to 
honour other directories. Can we get a simple toplevel method, but
even more importantly, can this also be documented? Ruby really needs
a better documentation; I am aware that many individual authors such
as jeremy or ioquatix write great documentation for their own projects,
but now that Google search also sucks (that's another topic though), 
finding useful information is getting harder and harder. Also, fewer people
using ruby means that this problem will just get worse, so we really need
to improve the documentation in general. Otherwise the decay effect will
just be amplified and people will follow the languages with better 
documentation (and more users, too).

TL;DR: Please specify a simple way for image-assets in different directories
without needing "use". Let's go back to oldschool API design rather than
DSLs-that-are-nice-to-read-but-not-flexible.
[on Mar 18, 2023](https://github.com/rack/rack/discussions/2061#discussioncomment-5351026)
Maintainer

**Comment by rubyFeedback** [comment]
> If you have any thoughts about simplifying the existing code

This is a chicken-egg problem because some other users may depend on this behaviour.

IMO it would be simpler to also have more toplevel-rack methods that allow such settings.
If you do not want to offer them automatically, then it could just be a special require
such as:

    require 'rack/configure' # or
    require 'rack/extended_configure' # or something like that

And then we have a more traditional API too, such as:

    Rack.image_directories = ['/tmp/img/cute_cats/','/home/joe/images/i_like_dogs']

and so forth.

Right now we seem to be required to think in terms of this DSL, such as:

    use Rack::Static,
      :urls => [
        '/home/pweb/cascading_style_sheets/'
      ]

By the way, it would also be nice if rack itself could have an extensive
short documentation aka a FAQ. Sinatra has this to some extent:

https://sinatrarb.com/one-oh-faq

Could be much longer and more extensive but at the least it has one.

I actually wanted to move away from sinatra since its documentation
is bad, but rack's documentation is even worse. :(
